### ISSUE TITLE
Decouple `core.StateDiff` from the feeder gateway and include Classes in `core.StateDiff` instead of passing them separately

### ISSUE THREAD

**Issue by IronGauntlets** [NethermindEth/juno#1014]
The [clean architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html) which we decided to use for Juno states the `Entities` (`core` package) should be independent of any influences from `Usecases` and be a stand-alone representation of a core concept. The reason for this is to have different layers in the project making it easier to maintain over time, where the core/entities are rarely changed.

[`core.StateDiff`](https://github.com/NethermindEth/juno/blob/c48883a31fa3c2e2b48d40385ea1f27705f7986e/core/state_update.go#L12-L19) contains fields `DeclaredV0Classes` and `DeclaredV1Classes` defined as `[]*felt.Felt` and `[]DeclaredV1Class` respectively. These fields have 3 usages:
1. Make it easy to convert between feeder gateway and RPC types.
2. Fetch full classes from the feeder gateway
3. Keep track of which classes to remove from persistent storage in case of revert.
4. (`DeclaredV1Classes` only) Update declared classes trie.

The 1st point above leads to the coupling of `core.StateDiff` to feeders and RPCs. In future if there is a change to the representation in the feeder or RPC, or if a new use case is introduced that impacts `core.StateDiff`, then the change will have a cascading impact throughout all of the different layers, defeating the purpose of clean architecture. Therefore, it is important to move away from the convenience of feeder/rpc's representations and shift the work to adapters for converting between structs.

Juno is required to store classes in persistent storage and since the current `core.StatDiff` doesn't contain classes we have to pass a `declaredClasses` map separately to `core.State.Update()`  and introduce another struct [`DeclaredClass`](https://github.com/NethermindEth/juno/blob/b47c0265420aa6a91f755d2062972b0dac6fdbb4/core/state.go#L316-L319) to keep track of which block number the class was declared on.

To address the issues mentioned above I propose to modify `core.StateDiff` as follows:

```go
type StateDiff struct {
	StorageDiffs      map[felt.Felt][]StorageDiff
	Nonces            map[felt.Felt]*felt.Felt
	DeployedContracts []DeployedContract
	DeclaredV0Classes []DeclaredClass
	DeclaredV1Classes []DeclaredClass
	ReplacedClasses   []ReplacedClass
}
```

As `DeclaredClass` contains the block number on which the class was declared and the `Class` interface which can be used to retrieve `ClassHash`, `CompiledClassHash` and the `Program`, it can replace all usage of the previous `DeclaredV0Classes` and `DeclaredV1Classes` (therefore, `DeclaredV1Class` struct should be removed). `declaredClasses` map would no longer need to be passed to `core.State.Update()`.

The proposed change would mean more work for the feeder and RPC adapters, as the feeder would have to fetch the classes and rpc adapter would have to convert`DeclaredClass` to respective fields in the `rpc.StateUpdate.StateDiff`. However, I believe for Juno's long-term maintainability these changes are necessary.

Please note this refactor should have no impact on the storage of classes and contracts.

**Comment by omerfirmak** [comment]
> include Classes in core.StateDiff instead of passing them separately

Will we store classes together with the `StateDiff`?

**Comment by IronGauntlets** [comment]
> Will we store classes together with the StateDiff?
Not necessarily. I will provide a better description of what I mean later today.

**Comment by omerfirmak** [comment]
> Please note this refactor should have no impact on the storage of classes and contracts.
We use `core` types as our storage types as well (which I believe makes sense), so how do we achieve this?

**Comment by IronGauntlets** [comment]
As far as I could tell, we don't store `StateDiff` as is in the db, each field of the State diff is stored separately (I may have missed something), so that is why I thought this refactor should have no impact on the anything we store.

**Comment by omerfirmak** [comment]
> As far as I could tell, we don't store StateDiff as is in the db

We actually do here, as a part of `core.StateUpdate`
https://github.com/NethermindEth/juno/blob/main/blockchain/blockchain.go#L571-L580
